### PR TITLE
[Dependency Scanning] Implement parallel imported module resolution

### PR DESCRIPTION
**Pull request by artemcm**
- **[Dependency Scanning] Refactor primary scan operations into `ModuleDependencyScanner` class**
From being a scattered collection of `static` methods in ScanDependencies.cpp
and member methods of `ASTContext`. This makes `ScanDependencies.cpp` much easier
to read, and abstracts the actual scanning logic away to a place with common
state which will make it easier to reason about in the future.

- **[Dependency Scanning] Implement parallel imported module resolution**
`ModuleDependencyScanner` maintains a Thread Pool along with a pool of workers
which are capable of executing a filesystem lookup of a named module dependency.
When resolving imports of a given Swift module, each import's resolution
operation can be issued asynchronously.

For the time being, the scanner is restricted to still run single-threaded, with a new frontend option `-parallel-scan` enabling multi-threaded dependency scanning.

### CODE REVIEW

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test Windows platform

**Comment by artemcm** [comment]
Windows build is failing to build the refactor because we're not linking something in correctly when building `SwiftIDETests`. 🤔 
```
swiftSerialization.lib(ModuleDependencyScanner.cpp.obj) : error LNK2019: unresolved external symbol "public: bool __cdecl swift::FrontendOptions::shouldTrackSystemDependencies(void)const " (?shouldTrackSystemDependencies@FrontendOptions@swift@@QEBA_NXZ) referenced in function "public: __cdecl swift::ModuleDependencyScanningWorker::ModuleDependencyScanningWorker(class swift::SwiftDependencyScanningService &,class swift::CompilerInvocation const &,class swift::SILOptions const &,class swift::ASTContext &,class swift::DependencyTracker &,class swift::DiagnosticEngine &)" (??0ModuleDependencyScanningWorker@swift@@QEAA@AEAVSwiftDependencyScanningService@1@AEBVCompilerInvocation@1@AEBVSILOptions@1@AEAVASTContext@1@AEAVDependencyTracker@1@AEAVDiagnosticEngine@1@@Z)
```

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test

File: test/ScanDependencies/batch_prescan.swift [hunk 1]
```diff
@@ -1,22 +0,0 @@
...
-
-// Check the contents of the JSON output
-// RUN: %validate-json %t/outputs/F.swiftmodule.json | %FileCheck %s -check-prefix=CHECK-SWIFT
-
-// CHECK-SWIFT: {
-// CHECK-SWIFT-NEXT:"imports": [
-// CHECK-SWIFT-NEXT:  "Swift",
-// CHECK-SWIFT-NEXT:  "F",
-// CHECK-SWIFT-NEXT:  "SwiftOnoneSupport"
-// CHECK-SWIFT-NEXT:]
```
Comment (xymus) [inline test/ScanDependencies/batch_prescan.swift:22]: Is it intentional to just delete this test?

File: include/swift/DependencyScan/ModuleDependencyScanner.h [hunk 2]
```diff
@@ -0,0 +1,150 @@
+//===--- ModuleDependencyScanner.h - Import Swift modules --------*- C++
+//-*-===//
```
Comment (xymus) [inline include/swift/DependencyScan/ModuleDependencyScanner.h:2]: 👀

**Review by xymus — APPROVED** [review]
LGTM!

File: test/ScanDependencies/batch_prescan.swift (hunk 1 above)
Comment (artemcm) [reply test/ScanDependencies/batch_prescan.swift:22]: Yeah, this functionality was not used anywhere at all and I do not foresee it being useful.

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test

**Pull request merged**